java/fury-core/src/main/java/org/apache/fury/meta/MetaStringEncoder.java

```diff
@@ -0,0 +1,288 @@
+    switch (encoding) {
+      case LOWER_SPECIAL:
+        return new MetaString(
+            input, encoding, specialChar1, specialChar2, encodeLowerSpecial(input), length * 5);
+      case LOWER_UPPER_DIGIT_SPECIAL:
+        return new MetaString(
+            input,
+            encoding,
+            specialChar1,
+            specialChar2,
+            encodeLowerUpperDigitSpecial(input),
+            length * 6);
+      case FIRST_TO_LOWER_SPECIAL:
+        return new MetaString(
+            input,
+            encoding,
+            specialChar1,
+            specialChar2,
+            encodeFirstToLowerSpecial(input),
+            length * 5);
+      case ALL_TO_LOWER_SPECIAL:
+        char[] chars = input.toCharArray();
+        int upperCount = countUppers(chars);
+        return new MetaString(
+            input,
+            encoding,
+            specialChar1,
+            specialChar2,
+            encodeAllToLowerSpecial(chars, upperCount),
+            (upperCount + length) * 5);
+      default:
+        byte[] bytes = input.getBytes(StandardCharsets.UTF_8);
+        return new MetaString(
+            input, Encoding.UTF_8, specialChar1, specialChar2, bytes, bytes.length * 8);
+    }
+  }
+
+  public Encoding computeEncoding(String input) {
+    if (input.isEmpty()) {
+      return Encoding.LOWER_SPECIAL;
+    }
+    char[] chars = input.toCharArray();
+    if (canBeLowerSpecialEncoded(chars)) {
+      return Encoding.LOWER_SPECIAL;
+    } else if (canBeLowerUpperDigitSpecialEncoded(chars)) {
+      if (countDigits(chars) != 0) {
+        return Encoding.LOWER_UPPER_DIGIT_SPECIAL;
+      } else {
+        int upperCount = countUppers(chars);
+        if (upperCount == 1 && Character.isUpperCase(chars[0])) {
+          return Encoding.FIRST_TO_LOWER_SPECIAL;
+        }
+        if ((chars.length + upperCount) * 5 < (chars.length * 6)) {
+          return Encoding.ALL_TO_LOWER_SPECIAL;
+        } else {
+          return Encoding.LOWER_UPPER_DIGIT_SPECIAL;
+        }
+      }
+    }
+    return Encoding.UTF_8;
+  }
+
+  private boolean canBeLowerSpecialEncoded(String input) {
+    return canBeLowerSpecialEncoded(input.toCharArray());
+  }
+
+  private boolean canBeLowerSpecialEncoded(char[] chars) {
+    for (char c : chars) {
+      if (c >= 'a' && c <= 'z') {
+        continue;
+      }
+      if (c == '.' || c == '_' || c == '$' || c == '|') {
+        continue;
+      }
+      // Character outside of LOWER_SPECIAL set
+      return false;
+    }
+    return true;
+  }
+
+  private int countDigits(char[] chars) {
+    int count = 0;
+    for (char c : chars) {
+      if (Character.isDigit(c)) {
+        count++;
+      }
+    }
+    return count;
+  }
+
+  private int countUppers(char[] chars) {
+    int upperCount = 0;
+    for (char c : chars) {
+      if (Character.isUpperCase(c)) {
+        upperCount++;
+      }
+    }
+    return upperCount;
+  }
+
+  private boolean canBeLowerUpperDigitSpecialEncoded(String input) {
```

Comment:
This function is not used, we should remove it.

Reply:
removed